Create integratord migration guide to 5.x (#36750)
wazuh/wazuh: docs/guide/migration/integratord-notifications.md

# Migrating Integratord to Dashboard Notifications

In Wazuh 4.x, third-party integrations were configured directly in `ossec.conf` on the manager using `<integration>` blocks, handled by the `integratord` daemon.

Wazuh 5.x replaces this approach with two dashboard plugins:

- **Notifications** — manages shared channels and connection settings for third-party services.
- **Alerting** — creates monitors that scan data and send messages through those channels.

All integration configuration is now done from the dashboard, with no changes required to `ossec.conf`.

---

> > **⚠️ Warning:** The Notifications plugin does not support bi-directional integrations such as Maltiverse or Virustotal, which were previously supported by `integratord`. These threat enrichment integrations are now handled through `security-analytics enrichments`. See [enrichment integrations](#enrichment-integrations).

Reviewer: ```suggestion
> **⚠️ Warning:** The Notifications plugin does not support bi-directional integrations such as Maltiverse or Virustotal, which were previously supported by `integratord`. These threat enrichment integrations are now handled through [Security Analytics enrichments](#enrichment-integrations).
```

> **Note:** This migration must be performed manually. There is no automatic tool to convert `<integration>` blocks from `ossec.conf` to the Wazuh 5.x dashboard configuration.

## Configuration mapping 4.x → 5.x

The following tables map each `<integration>` field from Wazuh 4.x to its equivalent location in the 5.x dashboard.

### Required configuration → Notifications channel

Configure these fields at **Explore > Notifications > Channel**.

| 4.x field        | 5.x field    | Notes                        |
| ---------------- | ------------ | ---------------------------- |
| `<name>`         | Channel name | Identifies the channel       |

> | `<hook_url>`     | Webhook URL  | Endpoint for the integration |

Reviewer: **suggestion**: consider mentioning there is a separate notification type for Slack that could be used. This could be created with `webhook` type too.

Author: In `Migration steps` > `1.2.1 Default channel type` mentions that there is some default channel types for services like Slack and Chime. I will expand section `1.2.2 custom webhook channel type` to explain that these services can be configured using the custom type too.

Author: Changed in: [1646ce1](https://github.com/wazuh/wazuh/pull/36750/commits/1646ce17d0287da7522d2d53a3176b3cd3748315)

> | `<api_key>`      | Headers      | Passed as a request header   |

Reviewer: **nitpick**: in a generic context, the API key or login credential depends on the external service, this could be the headers or payload/message.

Author: Changed in: [1646ce1](https://github.com/wazuh/wazuh/pull/36750/commits/1646ce17d0287da7522d2d53a3176b3cd3748315)

> | `<alert_format>` | Headers      | Passed as a request header   |

Reviewer: **nitpick**: the `alert_format` defined the format of the alert to "send" to the integration script. I think the equivalent could be the message/payload of the notification.

Author: At first i thought the alert_format was a reference of how the message was sent to the external service, but after looking more into it i was wrong.

The message/payload of the notification wouldn't be a good match either for these reasons:

First, the `<alert_format>` only has one possible value (json). In 4.x, the engine created temporal files using the alerts that matched with the `<integration>` block filters, and then the script read those files to create the payload and message. 

In all the default examples the `<alert_format>` was always present because json is better to parse into values than plain text, so that the python script could directly load the json, meaning that if the script expected json and found plain text it would crash.

I think `<alert_format>` does not have a direct match and can be considered deprecated, because the alerting plugin works with documents (json).

| `<rule_id>`        | Monitor query            | Monitors use queries on index patterns instead of rule matching. The matching field is `wazuh.rule.id`.                          |
| `<level>`          | Monitor query            | In 4.x the tag was a threshold, but in 5.x it's a value between `{low, medium, high}`. The matching field is `wazuh.rule.level`. |
| `<group>`          | Monitor query            | The group tag referred to the internal Wazuh component that generated the data. Now, its represented by `wazuh.integration.name` |
| `<event_location>` | Monitor query            | The 4.x field matched Wazuh modules or log sources via regex; in 5.x, queries using fields like `log.file.value`                 |
| `<options>`        | Trigger → Action message | See below                                                                                                                        |

![Configuration mapping channels](../../images/integratord-notifications/configuration_mapping.png)

#### From `<options>` to trigger actions

> In 4.x, the `<integration>` block invoked a built-in script (referenced by `<name>`) that sent a default JSON payload to the external service. The `<options>` block allowed appending extra fields to that payload.

Reviewer: **question**: What do you mean by `sent a default JSON payload to the external service`? If you are referring to the JSON payload sent in a request, the integration script could not send any request depending on the integration logic.

The `options` setting was used to provide options to the script. The statement seems to refer to the JSON paylodad sent to the external service.

Author: The built-in-scripts in 4.x processed and transformed the data into `json` structures, so if you wanted to add more fields to the message that was being sent, you had 2 options, either modify the script directly (good if it was something for every use case) or use the `<options>` tag to append fields to the json structure  (good if it was something specific, not re-usable).

The default scripts in python used the `dict.update()` method to update the json it created using the `<options>` json (if there was any), this meant that it replaced or appended fields.

This was entirely dependant on the script, meaning that if you created a custom script or edited the default script you could use the `<options>` tag for other purposes.

What i was trying to say is that the scripts had a defined structure but could be molded in specific use-cases by using the `<options>` tag.

I will rephrase it to be more clear.

<img width="1084" height="81" alt="imagen" src="https://github.com/user-attachments/assets/d30d043a-2e56-44ea-ad93-0e7add82e8e0" />

From: https://documentation.wazuh.com/current/user-manual/manager/integration-with-external-apis.html#:~:text=options,-%3E%20overwrites

Author: Changes in: [1646ce1](https://github.com/wazuh/wazuh/pull/36750/commits/1646ce17d0287da7522d2d53a3176b3cd3748315)

        <hook_url>YOUR_JIRA_HOOK_URL</hook_url>
        <api_key>EMAIL:API_KEY</api_key>
        <alert_format>json</alert_format>
    </integration>

    <integration>
        <name>virustotal</name>
        <api_key>YOUR_API_KEY</api_key>
        <group>syscheck</group>
        <level>10</level>
        <alert_format>json</alert_format>
    </integration>

    <integration>
        <name>maltiverse</name>
        <hook_url>https://api.maltiverse.com</hook_url>
        <api_key>YOUR_API_KEY</api_key>
        <level>5</level>
        <alert_format>json</alert_format>
    </integration>

</ossec_config>
```

## Migration steps

The steps below show one possible way to migrate the 4.x reference configuration to Wazuh 5.x. Use them as a guide and adapt the configuration to your own needs.

### 1. Configure the Notifications channel

Navigate to **Explore > Notifications > Channels**.

#### 1.1 Using default channels

Wazuh ships a set of pre-configured channels that are muted by default. To use one, un-mute it and fill in the required credentials. Each channel includes a brief description with configuration instructions.

**Example — PagerDuty channel:**

1. Un-mute the channel.

> 2. Edit the channel and enter the `<api_key>` from your 4.x config in the `X-Routing-Key` field under **Webhook Headers**.

Reviewer: ```suggestion
2. Edit the channel and enter the `<api_key>` from your 4.x config in the `X-Routing-Key` field under **Webhook headers**.
```

<img width="600" height="186" alt="Image" src="https://github.com/user-attachments/assets/13d6908a-e003-4f96-8dd2-1e26d913ecd6" />

3. Optionally update the channel name and description.

**Example — Jira channel:**

1. Un-mute the channel.
2. Edit the channel and enter the `<hook_url>` from your 4.x config on the **Webhook URL**

> 3. In 4.x the `<api_key>` had the structure `email:token`, in 5.x encode the string in base64 and then in the **Webhook Headers** **Authentication** field set the value `Basic encodedString`.

Reviewer: ```suggestion
3. In 4.x the `<api_key>` had the structure `email:token`, in 5.x encode the string in base64 and then set the value `Basic encodedString` in the `Authentication` header under **Webhook headers**.
```

For any other custom integration from 4.x, create a new channel using one of the types below.

##### 1.2.1 Default channel type

Wazuh 5.x provides dedicated channel types for common services such as `Slack`, `Chime`, and `Email`. These simplify setup — for `Chime`, for example, only a webhook URL is required.

##### 1.2.2 Custom webhook channel type

For integrations without a dedicated channel type, use the **Custom webhook** type. This is also the underlying type used by the Wazuh-provided `PagerDuty`, `Jira`, and `Shuffle` channels.

Custom webhook channels support:

- Methods: `PUT`, `POST`, or `PATCH`
- Endpoint definition: webhook URL or a custom URL with query parameters
- Custom headers: e.g. `Authorization`, `Content-Type`

---

### 2. Create one or more monitors using the Alerting plugin

With channels configured, create monitors that query data and send messages through those channels.

Navigate to **Explore > Alerting > Create Monitor**.

#### 2.1 Basic configuration

> A monitor defines the index pattern, schedule, and query/trigger/action structure. Because a single monitor supports multiple queries, triggers, and actions, the basic configuration determines whether you need one monitor or several.

Reviewer: ```suggestion
A monitor defines the indexes, schedule, and query/trigger/action structure. Because a single monitor supports multiple queries, triggers, and actions, the basic configuration determines whether you need one monitor or several.
```

**Example — Security Monitor (reusable for PagerDuty, Slack, Jira, and Shuffle):**

| Field         | Value                        |
| ------------- | ---------------------------- |
| Monitor name  | Security Monitor             |
| Monitor type  | Per document monitor         |
| Schedule      | By interval — every 1 minute |
| Index pattern | `wazuh-findings-v5-security` |

#### 2.2 Queries

A single monitor can have multiple queries. Create one query for each filter condition you had in 4.x.

**Example query for PagerDuty:**

> In the PagerDuty example, the original `<level>14</level>` maps to `wazuh.rule.level` is `high` in 5.x.

Reviewer: **nitpick**: there is no direct conversion of the old rule level to the new values, consider mention to review the wazuh.rule.leve information of the similar finding in 5.x. This could be confusing, other levels lowers to 14 could map too.

If a note is added about this, consider add it to other similar places using the `wazuh.rule.level`.

Author: Changed in: [1646ce1](https://github.com/wazuh/wazuh/pull/36750/commits/1646ce17d0287da7522d2d53a3176b3cd3748315)

The following payload is an example to create a task:

```json
{
  "fields": {
    "project": {
      "key": "YOUR_PROJECT_SPACE_KEY"
    },
    "issuetype": {
      "id": "10003"
    },
    "summary": "FIM alert on [{{ctx.alerts.0.sample_documents.0._source.file.path}}]",
    "description": {
      "version": 1,
      "type": "doc",
      "content": [
        {
          "type": "paragraph",
          "content": [
            {
              "type": "text",
              "text": "- State: {{ctx.alerts.0.sample_documents.0._source.wazuh.rule.title}}\n- Rule ID: {{ctx.alerts.0.sample_documents.0._source.wazuh.rule.id}}\n- Alert level: {{ctx.alerts.0.sample_documents.0._source.wazuh.rule.level}}\n- Agent: {{ctx.alerts.0.sample_documents.0._source.wazuh.agent.id}} {{ctx.alerts.0.sample_documents.0._source.wazuh.agent.name}}"
            }
          ]
        }
      ]
    }
  }
}
```

![Jira - Created task](../../images/integratord-notifications/created_task_on_jira.png)

For building message payloads with dynamic variables, refer to:

- [Mustache templates](https://mustache.github.io/mustache.5.html)

> - [Monitor variables](https://docs.opensearch.org/latest/observing-your-data/alerting/monitors/)
> - [Trigger variables](https://docs.opensearch.org/latest/observing-your-data/alerting/triggers/)
> - [Action variables](https://docs.opensearch.org/latest/observing-your-data/alerting/actions/)

Reviewer: **suggestion**: version the links. If using the version, we could need to add some the changes to the bumper.

```suggestion
- [Monitor variables](https://docs.opensearch.org/3.6/observing-your-data/alerting/monitors/)
- [Trigger variables](https://docs.opensearch.org/3.6/observing-your-data/alerting/triggers/)
- [Action variables](https://docs.opensearch.org/3.6/observing-your-data/alerting/actions/)
```

The reference to OpenSearch docs must be validated by @asteriscos.

Author: Leaving this untouched for now until further instructions

> ## Enrichment integrations

Reviewer: **suggestion**: provide information how new encrichments can be done.

Author: Enrichment process is managed by the engine in 5.x, and in the Dashboard Security Analytics plugin the user can add enrichment types to the different spaces, but there is no tool for the user to add new Enrichments.

Changed section in: [6d9381d](https://github.com/wazuh/wazuh/pull/36750/commits/6d9381dba2fbf03f6cfc3d64783cd38ad495236b)

The `integratord` daemon supported integrations like `Maltiverse` and `Virustotal`, these integrations received the alert that matched the `<integration>` block and after processing the alert, it was returned to wazuh with enriched information, creating a new alert in the process.

> The **Notifications** and **Alerting** plugins do not support this workflow directly (simply using channels and monitors), but the new **Security Analytics** plugin manages the creation of `Integrations`, `Decoders`, ... and the user can configure a series of pre-set enrichments for the custom spaces, meaning that the enrichment integrations are now integrated with the **Security Analytics** plugin. See the `Wazuh Dashboard > Modules > Security Analytics` documentation to create these integrations in 5.x.

Reviewer: ```suggestion
The **Notifications** and **Alerting** plugins do not support this workflow directly (simply using channels and monitors), but the new **Security Analytics** plugin manages the creation of `Integrations`, `Decoders`, ... and the user can configure a series of pre-set enrichments for the custom spaces, meaning that the enrichment integrations are now integrated with the **Security Analytics** plugin. See the `Wazuh dashboard > Modules > Security Analytics` documentation to create these integrations in 5.x.
```

2. Set **Action name** to `PagerDuty Incident`.
3. Set **Channel** to `PagerDuty Channel`.
4. Set the **Message** to the payload below.
5. Send a test message — you should see a test incident appear in your PagerDuty account.

> **⚠️ Warning:** The message payload must match the format expected by the receiving integration. If the format is incorrect, no event will appear on the external side. Consult each integration's API documentation before sending.

The following payload replicates what the default 4.x PagerDuty script sent:

```json
{
  "event_action": "trigger",
  "payload": {
    "summary": "Wazuh Alert: {{ctx.alerts.0.sample_documents.0._source.event.original}}",
    "source": "{{ctx.alerts.0.sample_documents.0._source.wazuh.agent.name}}",
    "severity": "critical",
    "custom_details": {
      "Rule ID": "{{ctx.alerts.0.sample_documents.0._source.wazuh.rule.id}}",
      "Rule Level": "{{ctx.alerts.0.sample_documents.0._source.wazuh.rule.level}}",
      "Agent OS": "{{ctx.alerts.0.sample_documents.0._source.wazuh.agent.host.os.name}}",
      "File Path": "{{ctx.alerts.0.sample_documents.0._source.file.path}}",
      "Monitor Name": "{{ctx.monitor.name}}"
    }
  }
}
```

> 

Reviewer 2: The warning in section 1.2.2 explains that Wazuh maps the X-Routing-Key header to the routing_key payload field automatically, so the field doesn't need to appear in the message. That said, this behavior isn't immediately obvious when reading the action payload example in isolation, a reader might not connect the header defined in step 1 with the payload sent in step 2. Would it be worth adding a short inline note near the payload example referencing this mechanism, so users don't have to cross-reference back to the warning?
Suggested note:
```suggestion
> **Note:** The routing_keyfield required by PagerDuty is injected automatically from theX-Routing-Key header configured in the channel — you do not need to add it to the payload manually.
```

Reviewer: Fixed https://github.com/wazuh/wazuh/pull/36750/commits/d90be50b53db391456274c7921ff8b8b9a09d0ef
